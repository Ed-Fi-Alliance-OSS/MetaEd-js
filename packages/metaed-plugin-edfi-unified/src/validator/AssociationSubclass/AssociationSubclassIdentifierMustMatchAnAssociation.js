// @flow
import type { MetaEdEnvironment } from '../../../../../packages/metaed-core/src/MetaEdEnvironment';
import type { ValidationFailure } from '../../../../../packages/metaed-core/src/validator/ValidationFailure';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.associationSubclass.forEach(associationSubclass => {
    if (!metaEd.entity.association.has(associationSubclass.baseEntityName)) {
      failures.push({
        validatorName: 'AssociationSubclassIdentifierMustMatchAnAssociation',
        category: 'error',
        message: `Association ${associationSubclass.metaEdName} based on ${associationSubclass.baseEntityName} does not match any declared Association.`,
        sourceMap: associationSubclass.sourceMap.type,
        fileMap: null,
      });
    }
  });
  return failures;
}
