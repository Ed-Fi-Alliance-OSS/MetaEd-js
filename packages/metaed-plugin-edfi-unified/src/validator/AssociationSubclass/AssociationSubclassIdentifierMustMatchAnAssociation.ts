import { MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.associationSubclass.forEach(entity => {
      if (
        getEntityFromNamespaceChain(
          entity.baseEntityName,
          entity.baseEntityNamespaceName,
          entity.namespace,
          'association',
        ) == null
      ) {
        failures.push({
          validatorName: 'AssociationSubclassIdentifierMustMatchAnAssociation',
          category: 'error',
          message: `Association ${entity.metaEdName} based on ${entity.baseEntityName} does not match any declared Association in namespace ${entity.baseEntityNamespaceName}.`,
          sourceMap: entity.sourceMap.metaEdName,
          fileMap: null,
        });
      }
    });
  });
  return failures;
}
