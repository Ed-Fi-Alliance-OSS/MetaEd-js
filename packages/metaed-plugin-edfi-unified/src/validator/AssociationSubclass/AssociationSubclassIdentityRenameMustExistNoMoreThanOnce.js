// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../metaed-core/index';
import { failSubclassIdentityRenamingMoreThanOnce } from '../ValidatorShared/FailSubclassIdentityRenamingMoreThanOnce';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.associationSubclass.forEach(associationSubclass => {
    failSubclassIdentityRenamingMoreThanOnce('AssociationSubclassIdentityRenameMustExistNoMoreThanOnce', associationSubclass, failures);
  });

  return failures;
}
