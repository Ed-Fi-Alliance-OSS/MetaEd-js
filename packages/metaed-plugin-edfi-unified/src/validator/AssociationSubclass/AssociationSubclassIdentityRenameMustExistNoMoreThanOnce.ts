import { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { failSubclassIdentityRenamingMoreThanOnce } from '../ValidatorShared/FailSubclassIdentityRenamingMoreThanOnce';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach(namespace => {
    namespace.entity.associationSubclass.forEach(associationSubclass => {
      failSubclassIdentityRenamingMoreThanOnce(
        'AssociationSubclassIdentityRenameMustExistNoMoreThanOnce',
        associationSubclass,
        failures,
      );
    });
  });

  return failures;
}
