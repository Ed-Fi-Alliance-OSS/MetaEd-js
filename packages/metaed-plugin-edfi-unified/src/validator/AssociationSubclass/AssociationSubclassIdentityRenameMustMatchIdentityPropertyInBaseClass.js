// @flow
import type { Association, MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { failSubclassIdentityRenameNotMatchingBaseClassIdentityProperty } from '../ValidatorShared/FailSubclassIdentityRenameNotMatchingBaseClassIdentityProperty';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.entity.associationSubclass.forEach(associationSubclass => {
    const extendedEntity: Association | void = metaEd.entity.association.get(associationSubclass.baseEntityName);
    failSubclassIdentityRenameNotMatchingBaseClassIdentityProperty(
      'AssociationSubclassIdentityRenameMustMatchIdentityPropertyInBaseClass',
      associationSubclass,
      extendedEntity,
      failures,
    );
  });

  return failures;
}
