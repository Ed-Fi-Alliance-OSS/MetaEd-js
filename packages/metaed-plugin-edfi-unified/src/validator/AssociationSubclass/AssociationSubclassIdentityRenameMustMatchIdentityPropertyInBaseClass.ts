import { Association, MetaEdEnvironment, ValidationFailure, Namespace } from '@edfi/metaed-core';
import { getEntityFromNamespaceChain } from '@edfi/metaed-core';
import { failSubclassIdentityRenameNotMatchingBaseClassIdentityProperty } from '../ValidatorShared/FailSubclassIdentityRenameNotMatchingBaseClassIdentityProperty';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.associationSubclass.forEach((associationSubclass) => {
      const extendedEntity: Association | null = getEntityFromNamespaceChain(
        associationSubclass.baseEntityName,
        associationSubclass.baseEntityNamespaceName,
        associationSubclass.namespace,
        'association',
      ) as Association | null;

      failSubclassIdentityRenameNotMatchingBaseClassIdentityProperty(
        'AssociationSubclassIdentityRenameMustMatchIdentityPropertyInBaseClass',
        associationSubclass,
        extendedEntity,
        failures,
      );
    });
  });
  return failures;
}
