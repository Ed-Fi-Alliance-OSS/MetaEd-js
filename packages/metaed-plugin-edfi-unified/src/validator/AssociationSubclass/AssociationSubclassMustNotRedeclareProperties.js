// @flow
import type { Association, AssociationExtension, MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';
import { failExtensionPropertyRedeclarations } from '../ValidatorShared/FailExtensionPropertyRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.associationSubclass.forEach(associationSubclass => {
    const extendedEntity: Association | AssociationExtension | void =
      metaEd.entity.association.get(associationSubclass.baseEntityName) || metaEd.entity.associationExtension.get(associationSubclass.baseEntityName);
    if (!extendedEntity) return;
    failExtensionPropertyRedeclarations('AssociationSubClassMustNotRedeclareProperties', associationSubclass, extendedEntity, failures);
  });

  return failures;
}
