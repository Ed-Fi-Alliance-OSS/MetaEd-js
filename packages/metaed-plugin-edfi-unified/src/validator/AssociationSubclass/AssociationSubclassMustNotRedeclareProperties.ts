import { Association, MetaEdEnvironment, ValidationFailure, Namespace } from '@edfi/metaed-core';
import { getEntityFromNamespaceChain } from '@edfi/metaed-core';
import { failExtensionPropertyRedeclarations } from '../ValidatorShared/FailExtensionPropertyRedeclarations';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.associationSubclass.forEach((associationSubclass) => {
      const extendedEntity: Association | null = getEntityFromNamespaceChain(
        associationSubclass.baseEntityName,
        associationSubclass.baseEntityNamespaceName,
        associationSubclass.namespace,
        'association',
      ) as Association | null;

      if (!extendedEntity) return;
      failExtensionPropertyRedeclarations(
        'AssociationSubClassMustNotRedeclareProperties',
        associationSubclass,
        extendedEntity,
        failures,
      );
    });
  });
  return failures;
}
