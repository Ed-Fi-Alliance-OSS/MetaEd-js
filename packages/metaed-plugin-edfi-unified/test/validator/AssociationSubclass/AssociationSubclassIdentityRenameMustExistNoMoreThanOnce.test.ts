import {
  newMetaEdEnvironment,
  MetaEdTextBuilder,
  AssociationBuilder,
  AssociationSubclassBuilder,
  NamespaceBuilder,
} from '@edfi/metaed-core';
import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';
import { validate } from '../../../src/validator/AssociationSubclass/AssociationSubclassIdentityRenameMustExistNoMoreThanOnce';

describe('when association subclass renames base identity more than once', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName = 'EntityName';
  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartAssociation(entityName)
      .withDocumentation('EntityDocumentation')
      .withAssociationDomainEntityProperty('PropertyName1', 'PropertyDocumentation')
      .withAssociationDomainEntityProperty('PropertyName2', 'PropertyDocumentation')
      .withStringIdentity('PropertyName3', 'PropertyDocumentation3', '100')
      .withStringIdentity('PropertyName4', 'PropertyDocumentation4', '100')
      .withEndAssociation()

      .withStartAssociationSubclass('SubclassName', entityName)
      .withDocumentation('EntityDocumentation')
      .withStringIdentityRename('PropertyName5', 'PropertyName3', 'PropertyDocumentation', '100')
      .withStringIdentityRename('PropertyName6', 'PropertyName4', 'PropertyDocumentation', '100')
      .withEndAssociationSubclass()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new AssociationBuilder(metaEd, []))
      .sendToListener(new AssociationSubclassBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one association', (): void => {
    expect(coreNamespace.entity.association.size).toBe(1);
  });

  it('should build one associationSubclass', (): void => {
    expect(coreNamespace.entity.associationSubclass.size).toBe(1);
  });

  it('should have validation failures', (): void => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('AssociationSubclassIdentityRenameMustExistNoMoreThanOnce');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot(
      'when association subclass renames base identity more than once should have validation failure -> message',
    );
    expect(failures[0].sourceMap).toMatchSnapshot(
      'when association subclass renames base identity more than once should have validation failure -> sourceMap',
    );
    expect(failures[1].validatorName).toBe('AssociationSubclassIdentityRenameMustExistNoMoreThanOnce');
    expect(failures[1].category).toBe('error');
    expect(failures[1].message).toMatchSnapshot(
      'when association subclass renames base identity more than once should have validation failure -> message',
    );
    expect(failures[1].sourceMap).toMatchSnapshot(
      'when association subclass renames base identity more than once should have validation failure -> sourceMap',
    );
  });
});
