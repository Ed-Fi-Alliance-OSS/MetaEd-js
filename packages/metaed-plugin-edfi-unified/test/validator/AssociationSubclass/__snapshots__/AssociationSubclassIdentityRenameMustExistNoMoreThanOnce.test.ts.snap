// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`when association subclass renames base identity more than once should have validation failures: when association subclass renames base identity more than once should have validation failure -> message 1`] = `"Association Subclass SubclassName based on EntityName tries to rename properties PropertyName3, PropertyName4. Only one identity rename is allowed for a given Association Subclass."`;

exports[`when association subclass renames base identity more than once should have validation failures: when association subclass renames base identity more than once should have validation failure -> message 2`] = `"Association Subclass SubclassName based on EntityName tries to rename properties PropertyName3, PropertyName4. Only one identity rename is allowed for a given Association Subclass."`;

exports[`when association subclass renames base identity more than once should have validation failures: when association subclass renames base identity more than once should have validation failure -> sourceMap 1`] = `
Object {
  "column": 32,
  "line": 27,
  "tokenText": "PropertyName3",
}
`;

exports[`when association subclass renames base identity more than once should have validation failures: when association subclass renames base identity more than once should have validation failure -> sourceMap 2`] = `
Object {
  "column": 32,
  "line": 32,
  "tokenText": "PropertyName4",
}
`;
